---
title: |
  | _Aedes aegypti_ mRNA expression --
  | Comparison with other studies
pagetitle: "Compare"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster, OSU"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
bibliography: compare.bib
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
## Install/load packages
if(! "pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse",        # Misc. data manipulation and plotting
              "here",             # Managing file paths
              "readxl",           # Reading Excel files
              "VennDiagram",      # Venn Diagram
              "RColorBrewer",     # Plotting colors
              "plotly",           # Interactive plots
              "glue",             # Pasting text
              "DT", "kableExtra") # Tables             
pacman::p_load(char = packages)
```

```{r}
## Source other scripts
source(here("scripts/report_funs.R"))
source(here("scripts/compare_funs.R"))
```

<br>

## Introductory notes

- Currently, the DEGs from our study included here are only those
  **from LRT at 3% sugar water,**
  **with the comparisons of Virgin vs. LVP at 6 hpm and at 24 hpm.**
  To keep the number of possible comparisons and combinations manageable,
  it may work best to keep our DEGs limited to one list in this way.
  But besides DEGs at 3%, that list could also contain DEGs at _either_
  3% or 10%, or DEGs at _both_ 3% and 10%. Feedback welcome...

- I am _not_ currently distinguishing between the direction of DE --
  i.e. whether genes are under- versus overexpressed in different treatments.

- All reported p-values have been corrected for multiple testing
  (raw p-values are not shown).

### Data sources

- **@alfonso-parra_mating-induced_2016** compared expression in reproductive tracts
  of virgin and mated females at 0, 6 and 24 hours post-mating.
  The data was taken from a Supplement of a different paper by the same lab
  (@amaro_seminal_2021),
  in which these authors re-analyzed the Alfonso-Parra data using a new
  _A. aegypti_ assembly.
  - The supplementary Excel file with the data is "Additional file 2" --
    [link to the file](https://static-content.springer.com/esm/art%3A10.1186%2Fs12864-021-08201-0/MediaObjects/12864_2021_8201_MOESM2_ESM.xlsx).
    This file only contains DEGs.
  - [Link to the paper online](https://journals.plos.org/plosntds/article?id=10.1371/journal.pntd.0004451).

- **@alonso_gene_2019** compared expression in head and thorax of virgin
  vs. inseminated females -- _**it is not clear to me how many hours post-mating**_.
  **This study quantified expression on a per-transcript basis.**
  For each gene, I only kept the results for the transcript with the lowest
  adjusted p-value, and then converted the transcript identifier to the
  parent gene identifier.
  - The supplementary Excel file with the data is "Supplementary Dataset 1" --
    [link to the file](https://static-content.springer.com/esm/art%3A10.1038%2Fs41598-019-52268-5/MediaObjects/41598_2019_52268_MOESM1_ESM.xlsx).
    This file contains both DEGs and non-DEGs.
  - [Link to the paper online](https://www.nature.com/articles/s41598-019-52268-5).

- **@amaro_seminal_2021** compared expression in the female LRT 24 h after injecting
  extract from male accessory glands and seminal vesicles (MAG extract).
  **DISCUSS GENE SELECTION**
  - The Supplementary file to the paper was a PDF, but the authors sent us an Excel file by email.
  - [Link to the paper online](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08201-0).

- **@camargo_mating_2020** compared expression in spermathecae from unfed virgins,
  mated only, and mated and blood-fed females at 6, 24, and 72 h post-mating.
  I only used the data from 6 and 24 hpm for non-blood-fed females.
  - The supplementary Excel file with the data is "Supplementary file3" --
    [link to the file](https://static-content.springer.com/esm/art%3A10.1038%2Fs41598-020-71904-z/MediaObjects/41598_2020_71904_MOESM3_ESM.xlsx).
    This file only contains DEGs.
  - [Link to the paper online](https://www.nature.com/articles/s41598-020-71904-z).
  
- **@pascini_transcriptional_2020** compared expression in spermathecae of
  virgin vs inseminated females, 7 days post-emergence --
  and ~7 days post-mating (?).
  This study performed a de novo transcriptome assembly, and later BLASTed
  their transcripts/genes to the _A. aegypti_ genome.
  As a result, some _A. aegypti_ gene IDs occur multiple times,
  and I kept only the gene with lowest adjusted p-value. 
  - The supplementary Excel file with the data is "Additional file 5" --
    [link to the file](https://ndownloader.figstatic.com/files/22090851).
    This file contains both DEGs and non-DEGs.
  - [Link to the paper online](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-020-6543-y).

```{r}
overview <- tribble(
  ~Author,       ~Year, ~Tissue,        ~hpm6, ~hpm24, ~`hpm - other`, ~Treatment,         ~`DEGs only?`,
  "Alfonso-Parra", 2016, "RT",            "yes", "yes", NA,          "mating",           "yes",
  "Alonso",        2019, "head & thorax", "no",  "no", "unknown!",   "mating",           "no",
  "Amaro",         2021, "LRT",           "no",  "yes", NA,          "MAG insemination", "yes",
  "Camargo",       2020, "spermathecae",  "yes", "yes", NA,          "mating",           "yes",
  "Pascini",       2020, "spermathecae",  "no",  "no", "7 days",     "mating",           "no"
) %>%
  mutate(Year = as.integer(Year))
```

<br>

#### Summary table

```{r}
overview %>% make_dt(simple_mode = TRUE)
```

### Abbreviations

- `hpm` -- Hours post-mating
- `LFC` -- Log2-fold change in mean expression level
- `DE` -- Differentially Expressed
- `DEG`-- Differentially Expressed Gene

<br>

-----

```{r}
## Set contrasts of interest
contrasts <- c("lvp06_vir06", "lvp24_vir24")
contrast_names <- c("LVP v. virgin (6 hpm)", "LVP v. virgin (24 hpm)")

## Define studies
study_levels <- c("ours", "alfpar", "alonso", "amaro", "camargo", "pascini")
studies_long <- c("This study", "Alfonso-Parra", "Alonso", "Amaro",
                  "Camargo", "Pascini")

studies_un <- c("ours", "alonso", "pascini")
studies_6hpm <- c("ours", "alfpar", "camargo")
studies_24hpm <- c("ours", "alfpar", "amaro", "camargo")
```

```{r}
## Define input files -- own DE results and gene lookup
indir_DE <- here("results", "DE")
DE_file <- here(indir_DE, "DE_all.txt")
gene_file <- here(indir_DE, "gene_info.txt")
other_studies_file <- here("results/other-studies/other-studies_combined.txt")
```

```{r}
## Read input files
gene_df <- read_tsv(gene_file, show_col_types = FALSE) %>%
  select(gene_id, description) %>%
  mutate(description = str_trunc(description, width = 40))

DE_res <- read_tsv(DE_file, show_col_types = FALSE) %>%
  filter(sugar == "sug03") %>% 
  mutate(tissue = sub(".*\\d(\\w+)", "\\1", treat_a),
         tissue = ifelse(sugar == "sug10", "lrt10", tissue),
         trt_a = sub("(.*\\d+)\\w+", "\\1", treat_a),
         trt_b = sub("(.*\\d+)\\w+", "\\1", treat_b),
         contrast = paste0(trt_a, "_", trt_b),
         hpm = as.integer(sub(".*[a-z](\\d+)", "\\1", trt_b))) %>%
  filter(contrast %in% contrasts, tissue == "lrt") %>%
  select(gene_id, lfc, padj, hpm) %>%
  mutate(study = "ours")

others <- read_tsv(other_studies_file, show_col_types = FALSE)
```

```{r}
## Subsets for Venn diagrams
DE6 <- DE_res %>% filter(hpm == 6, padj < 0.05)
DE24 <- DE_res %>% filter(hpm == 24, padj < 0.05)
alfpar_6h <- others %>% filter(study == "alfpar", hpm == 6)
alfpar_24h <- others %>% filter(study == "alfpar", hpm == 24)
camargo_6h <- others %>% filter(study == "camargo", hpm == 6)
camargo_24h <- others %>% filter(study == "camargo", hpm == 24)
pascini <- others %>% filter(study == "pascini", padj < 0.05)
alonso <- others %>% filter(study == "alonso", padj < 0.05)
amaro <- others %>% filter(study == "amaro", padj < 0.05)
```

```{r}
## Combine own DE results and those of the other studies
long_all <- bind_rows(DE_res, others) %>%
  mutate(hpm = as.integer(hpm))

## Wide format
wide_all <- long_all %>%
  mutate(hpm = ifelse(is.na(hpm), 6, hpm)) %>%    # Pascini & Alonso => 6 hpm
  bind_rows(pascini %>% mutate(hpm = 24)) %>%     # Copy of Pascini => 24 hpm
  bind_rows(alonso %>% mutate(hpm = 24)) %>%      # Copy of Alonso => 24 hpm
  mutate(study = factor(study, study_levels)) %>% 
  arrange(study) %>%
  #filter(gene_id == "AAEL000001", hpm == 6, study == "ours")
  pivot_wider(id_cols = c(gene_id, hpm),
              values_from = c(lfc, padj), names_from = study) %>%
  left_join(gene_df, by = "gene_id") %>%
  relocate(description, .after = gene_id) %>%
  relocate(contains("padj"), .after = hpm) %>% 
  arrange(gene_id)
```

```{r}
#   gene_id      lfc  padj   hpm study
#   <chr>      <dbl> <dbl> <dbl> <fct>
# 1 AAEL000001 0.086 0.986     6 ours 
# 2 AAEL000001 0.007 1         6 ours 
```

```{r}
## Create subsets of the wide df for tables and LFC plots
wide_6strict <- wide_all %>%         # 6 hpm only
  filter(hpm == 6) %>% 
  select(gene_id, description, hpm, contains(studies_6hpm))
  
wide_6un <- wide_all %>%             # 6 hpm ours vs other unknown
  filter(hpm == 6) %>% 
  select(gene_id, description, hpm, contains(c("ours", studies_un)))

wide_24strict <- wide_all %>%        # 24 hpm only
  filter(hpm == 24) %>% 
  select(gene_id, description, hpm, contains(studies_24hpm))
  
wide_24un <- wide_all %>%             # 24 hpm ours vs other unknown
  filter(hpm == 24) %>% 
  select(gene_id, description, hpm, contains(c("ours", studies_un)))
```

```{r}
## Create lists for Venn diagrams
list_6strict <- list(DE6$gene_id, alfpar_6h$gene_id, camargo_6h$gene_id)
names(list_6strict) <- c("This study", "Alfonso-Parra et al.", "Camargo et al.")

list_6un <- list(DE6$gene_id, alonso$gene_id, pascini$gene_id)
names(list_6un) <- c("This study", "Alonso et al.", "Pascini et al.")

list_24strict <- list(DE24$gene_id, alfpar_24h$gene_id, amaro$gene_id,
                      camargo_24h$gene_id)
names(list_24strict) <- c("This study", "Alfonso-Parra et al.",
                          "Amaro et al.", "Camargo et al.")

list_24un <- list(DE24$gene_id, alonso$gene_id, pascini$gene_id)
names(list_24un) <- c("This study", "Alonso et al.", "Pascini et al.")
```

```{r}
multsig_cap <- paste0("nr. signif. = The number of studies ",
                      "in which the gene is DE",
                      "; p = adjusted p-value; lfc = log2-fold change.")
```

```{r}
lfc_text <- paste(
  "\n\n- These show all available data for a given study --",
  "either LFC values for all genes or only for DEGs.",
  "If a gene's LFC is not available for one of the two studies in a plot,",
  "it is not currently plotted at all.",
  "Therefore, for a pairwise plot of two studies where LFC values are only",
  "available for DEGs will only show genes that are DE in both studies, etc.",
  "\n\n- The Pascini dataset has a lot of infinite LFC values due to having 0",
  "read counts in one group.")
```

```{r}
table_text <- paste(
  "\n\n- Only genes that are DE in our study and at least one other study are listed.",
  "\n\n- Cells with a green background are significant (adjusted p-value < 0.05)."
)
```

## Comparisons at 6 hpm

### Only 6 hpm {.tabset .tabset-fade .tabset-pills}

#### Venn

```{r, results="hide"}
flist <- list_6strict
p <- venn.diagram(x = flist, fill = brewer.pal(length(flist), "Set2"),
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

#### Table

```{r, results='asis'}
cat(table_text)
```

<br>

```{r}
multsig_dt(wide_6strict)
```

#### Interactive LFC plots

```{r}
combs <- combn(studies_6hpm, 2)
```

```{r, results='asis'}
cat(lfc_text)
```

```{r}
l <- htmltools::tagList()
for (i in 1:ncol(combs)) {
  l[[i]]<- plot_lfc(wide_6strict, combs[1, i], combs[2, i], interactive = TRUE)
}
l
```

#### Non-interactive LFC plots

```{r, results='asis'}
cat(lfc_text)
```

```{r, out.width="90%"}
walk(1:ncol(combs), function(x) {
  plot_lfc(wide_6strict, combs[1, x], combs[2, x])
  })
```

### Our 6 hpm and other/unknown {.tabset .tabset-fade .tabset-pills}

#### Venn

```{r, results="hide"}
flist <- list_6un
p <- venn.diagram(x = flist, fill = brewer.pal(length(flist), "Set2"),
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

<br>

-----

#### Table

```{r, results='asis'}
cat(table_text)
```

<br>

```{r}
multsig_dt(wide_6un)
```

<br>

-----

#### Interactive LFC plots

```{r}
combs <- combn(studies_un, 2)
```

```{r, results='asis'}
cat(lfc_text)
```

```{r}
l <- htmltools::tagList()
for (i in 1:ncol(combs)) {
  l[[i]]<- plot_lfc(wide_6un, combs[1, i], combs[2, i], interactive = TRUE)
}
l
```

<br>

-----

#### Non-interactive LFC plots

```{r, results='asis'}
cat(lfc_text)
```

```{r, out.width="90%"}
walk(1:ncol(combs), function(x) {
  plot_lfc(wide_6un, combs[1, x], combs[2, x])
  })
```

<br>

-----

## Comparisons at 24 hpm

### Only 24 hpm {.tabset .tabset-fade .tabset-pills}

#### Venn

```{r, results="hide"}
flist <- list_24strict
p <- venn.diagram(x = flist, fill = brewer.pal(length(flist), "Set2"),
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

#### Table

```{r, results='asis'}
cat(table_text)
```

<br>

```{r}
multsig_dt(wide_24strict)
```

#### Interactive LFC plots

```{r}
combs <- combn(studies_24hpm, 2)
```

```{r, results='asis'}
cat(lfc_text)
```

```{r}
l <- htmltools::tagList()
for (i in 1:ncol(combs)) {
  l[[i]]<- plot_lfc(wide_24strict, combs[1, i], combs[2, i], interactive = TRUE)
}
l
```

#### Non-interactive LFC plots

```{r, results='asis'}
cat(lfc_text)
```

```{r, out.width="90%"}
walk(1:ncol(combs), function(x) {
  plot_lfc(wide_24strict, combs[1, x], combs[2, x])
  })
```

### Our 24 hpm and other/unknown {.tabset .tabset-fade .tabset-pills}

#### Venn

```{r, results="hide"}
flist <- list_24un
p <- venn.diagram(x = flist, fill = brewer.pal(length(flist), "Set2"),
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

#### Table

```{r, results='asis'}
cat(table_text)
```

<br>

```{r}
multsig_dt(wide_24un) 
```

<br>

-----

#### Interactive LFC plots

```{r}
combs <- combn(studies_un, 2)
```

```{r, results='asis'}
cat(lfc_text)
```

```{r}
l <- htmltools::tagList()
for (i in 1:ncol(combs)) {
  l[[i]]<- plot_lfc(wide_24un, combs[1, i], combs[2, i], interactive = TRUE)
}
l
```

<br>

-----

#### Non-interactive LFC plots

```{r, results='asis'}
cat(lfc_text)
```

```{r, out.width="90%"}
walk(1:ncol(combs), function(x) {
  plot_lfc(wide_24un, combs[1, x], combs[2, x])
  })
```

<br>

----

## References
