---
title: |
  | Differential expression downstream analyses
pagetitle: "DE"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
## Install/load packages
if(! "pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse",        # Misc. data manipulation and plotting
              "here",             # Managing file paths
              "VennDiagram",      # Venn Diagram
              "RColorBrewer",     # Plotting colors
              "DT", "kableExtra") # Tables             
pacman::p_load(char = packages)
```

```{r}
## Script with functions
source(here("scripts", "GO_fun.R"))
```

```{r}
## Set contrasts
contrasts_ab <- c("akh06ab_vir06ab", "akh24ab_vir24ab",
                  "lvp06ab_vir06ab", "lvp24ab_vir24ab")
                  #"lvp24ab_lvp06ab", "akh24ab_akh06ab")
contrasts_ht <- c("akh06ht_vir06ht", "akh24ht_vir24ht",
                  "lvp06ht_vir06ht", "lvp24ht_vir24ht")
                  #"lvp24ht_lvp06ht", "akh24ht_akh06ht")
contrasts_lrt <- c("akh06lrt_vir06lrt", "akh24lrt_vir24lrt",
                  "lvp06lrt_vir06lrt", "lvp24lrt_vir24lrt")
                  #"lvp24lrt_lvp06lrt", "akh24lrt_akh06lrt")
contrasts_alltiss <- c(contrasts_ab, contrasts_ht, contrasts_lrt)
```

```{r}
## Define input files
indir_DE <- here("results", "DE")
indir_GO <- here("results", "GO")

GO_res_file <- here(indir_GO, "GO_all.txt")
DE_res_file <- here(indir_DE, "DE_all.txt")
gene_file <- here(indir_DE, "gene_info.txt")
```

```{r}
## Read input files
DE_res <- read_tsv(DE_res_file, show_col_types = FALSE) %>%
  mutate(sig = ifelse(padj < 0.05, TRUE, FALSE))

GO_res <- read_tsv(GO_res_file, show_col_types = FALSE) %>%
  mutate(sig = ifelse(padj < 0.05 & numDEInCat > 1, TRUE, FALSE))
GO_sig <- GO_res %>% filter(padj < 0.05)

gene_df <- read_tsv(gene_file, show_col_types = FALSE) %>%
  select(gene_id, description) %>%
  mutate(description = str_trunc(description, width = 50))
```

```{r}
## Make a df with all pairwise contrasts
contrast_df <- DE_res %>%
  distinct(contrast, .keep_all = TRUE) %>% 
  separate(contrast, remove = FALSE,
           into = c("group_a", "group_b"), sep = "_") %>% 
  mutate(tissue = gsub(".*\\d(\\w+)", "\\1", group_a),
         time = gsub(".*(\\d\\d).*_.*(\\d\\d).*", "\\1_\\2", contrast),
         group_a = gsub("(.*\\d+)\\w+", "\\1", group_a),
         group_b = gsub("(.*\\d+)\\w+", "\\1", group_b)) %>%
  mutate(timepoint = factor(time, levels = c("06_06", "24_24", "24_06"))) %>%
  select(contrast, group_a, group_b, tissue, time)
```

```{r}
## Function to create a df with genes significant in multiple focal contrasts
multisig <- function(tissue, contrasts, DE_res, minsig = 3) {
  DE_res %>%
    filter(sig == TRUE,
           contrast %in% contrasts,
           grepl(paste0(tissue, "$"), contrast)) %>%
    add_count(gene_id, name = "n_sig") %>%
    filter(n_sig >= minsig) %>% 
    select(gene_id, n_sig, contrast, padj) %>%
    pivot_wider(id_cols = c(gene_id, n_sig),
                names_from = contrast, values_from = padj) %>%
    arrange(desc(n_sig), gene_id) %>%
    rename("sign. in n contrasts" = n_sig)
}
```
<br>

----

## Differentially expressed genes

### Nr of genes DE per pairwise contrast

```{r}
sig_by_comp <- DE_res %>%
  mutate(contrast = factor(contrast), sig = factor(sig)) %>% 
  group_by(contrast, sig, .drop = FALSE) %>%
  count(name = "nsig") %>%
  filter(sig == TRUE) %>% 
  ungroup() %>%
  select(contrast, nsig) %>%
  full_join(contrast_df, by = "contrast") %>% 
  select(group_a, group_b, tissue, time, nsig) %>%
  pivot_wider(id_cols = c(group_a, group_b, time),
              names_from = tissue, values_from = nsig) %>%
  arrange(time, group_a, group_b) %>%
  select(-time) %>%
  rename("group A" = group_a, "group B" = group_b,
         abdomen = ab, "head & thorax" = ht, LRT = lrt)

datatable(sig_by_comp, filter = "top", class = "compact",
          options = list(scrollX = TRUE, autoWidth = TRUE))
```

### Genes DE for multiple contrasts {.tabset .tabset-fade .tabset-pills}

#### Abdomen

```{r}
multisig_ab <- multisig("ab", contrasts_ab, DE_res, minsig = 2) %>%
  left_join(gene_df, by = "gene_id")

datatable(multisig_ab, filter = "top", class = "compact",
          options = list(scrollX = TRUE, autoWidth = TRUE)) %>%
  formatSignif(colnames(multisig_ab)[3:ncol(multisig_ab) - 1], digits = 3)
```

#### Head & thorax

```{r}
multisig_ht <- multisig("ht", contrasts_ht, DE_res, minsig = 2) %>%
  left_join(gene_df, by = "gene_id")

datatable(multisig_ht, filter = "top", class = "compact",
          options = list(scrollX = TRUE, autoWidth = TRUE)) %>%
  formatSignif(colnames(multisig_ht)[3:ncol(multisig_ht) - 1], digits = 3)
```

#### LRT

```{r}
multisig_lrt <- multisig("lrt", contrasts_lrt, DE_res, minsig = 2) %>%
  left_join(gene_df, by = "gene_id")

datatable(multisig_lrt, filter = "top", class = "compact",
          options = list(scrollX = TRUE, autoWidth = TRUE)) %>%
  formatSignif(colnames(multisig_lrt)[3:ncol(multisig_lrt) - 1], digits = 3)
```

### Venn diagrams {.tabset .tabset-fade .tabset-pills}

```{r}
## Function to get a vector with significantly DE genes for a pairwise contrast
get_genes <- function(my_contrast, DE_res) {
  DE_res %>%
    filter(sig == TRUE, contrast == my_contrast) %>%
    pull(gene_id)
}

## Set colors for Venn diagram
my_cols <- brewer.pal(4, "Pastel2")
```

#### Abdomen

```{r}
DE_genes_ab <- map(.x = contrasts_ab, .f = get_genes, DE_res)
names(DE_genes_ab) <- gsub("ab", "", contrasts_ab)

p <- venn.diagram(x = DE_genes_ab,
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  fill = my_cols, cat.cex = 0.8, cat.fontface = "bold")
grid::grid.draw(p)
```

<br>

----

#### Head & thorax

```{r}
DE_genes_ht <- map(.x = contrasts_ht, .f = get_genes, DE_res)
names(DE_genes_ht) <- gsub("ht", "", contrasts_ht)

p <- venn.diagram(x = DE_genes_ht,
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  fill = my_cols, cat.cex = 0.8, cat.fontface = "bold")
grid::grid.draw(p)
```

<br>

----

#### LRT

```{r}
DE_genes_lrt <- map(.x = contrasts_lrt, .f = get_genes, DE_res)
names(DE_genes_lrt) <- gsub("lrt", "", contrasts_lrt)

p <- venn.diagram(x = DE_genes_lrt,
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  fill = my_cols, cat.cex = 0.8, cat.fontface = "bold")
grid::grid.draw(p)
```

```{r}
# x <- DE_genes_lrt
# twoway <- combn(names(x), 2)
# run2way <- function(idx, x) intersect(x[[twoway[1, idx]]], x[[twoway[2, idx]]])
# lapply(1:ncol(twoway), run2way, x)
# names(DE_genes_lrt)
```

<br>

----

## Gene Ontology

### Methods

The R/Bioconductor package
[`GOseq`](https://bioconductor.org/packages/release/bioc/html/goseq.html)
was used to test for GO term enrichment among differentially expressed genes
for each pairwise contrast.
Gene Ontology assignments to genes were extracted from a VectorBase GAF
file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gaf/).
Gene lengths are used in the statistical model of `GOseq`,
and there were extracted from a VectorBase GFF file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gff/data/).
P-values were corrected for multiple testing using the Benjamini-Hochberg
correction.

### Tables {.tabset .tabset-fade .tabset-pills}

#### Nr significant GOs per contrast

```{r}
sig_by_comp <- GO_res %>%
  group_by(contrast, sig, .drop = FALSE) %>%
  count(name = "nsig") %>%
  filter(sig == TRUE) %>% 
  ungroup() %>%
  select(contrast, nsig) %>%
  full_join(contrast_df, by = "contrast") %>% 
  select(group_a, group_b, tissue, time, nsig) %>%
  pivot_wider(id_cols = c(group_a, group_b, time),
              names_from = tissue, values_from = nsig) %>%
  arrange(time, group_a, group_b) %>%
  select(-time) %>%
  rename("group A" = group_a, "group B" = group_b,
         abdomen = ab, "head & thorax" = ht, LRT = lrt)

datatable(sig_by_comp, filter = "top", class = "compact",
          options = list(scrollX = TRUE, autoWidth = TRUE))
```

#### Nr significant contrasts per GO term

```{r}
sig_by_term <- GO_sig %>%
  add_count(category) %>%
  group_by(category) %>%
  mutate("signif. contrasts" = paste(contrast, collapse = ", ")) %>%
  select("n significant" = n, category, ontology, description,
         `signif. contrasts`)

datatable(sig_by_term, class = "compact",
          filter = "top", options = list(scrollX = TRUE))
```

#### All significant GO terms

```{r}
sig_smr <- pivot_wider(GO_sig, id_cols = c(category, description),
                       names_from = contrast, values_from = padj)

numcols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]
datatable(sig_smr, class = "compact",
          filter = "top", options = list(scrollX = TRUE)) %>%
  formatSignif(numcols, digits = 3)
```

### Heatmaps {.tabset .tabset-fade .tabset-pills}

#### Abdomen

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, contrasts_ab) %>% goplot(ylabsize = 7)
```

#### Head & thorax

```{r, out.width="100%"}
prep_goplot(GO_res, contrasts_ht) %>% goplot()
```

#### LRT

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, contrasts_lrt) %>% goplot(ylabsize = 7)
```

#### All

```{r, fig.height = 8, out.width="100%"}
prep_goplot(GO_res, contrasts_alltiss) %>% goplot(ylabsize = 5)
```

