---
title: |
  | _Aedes aegypti_ mRNA expression --
  | Gene set enrichment analyses
pagetitle: "enrichment"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster, OSU"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
## Install/load packages
if(! "pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse",        # Misc. data manipulation and plotting
              "here",             # Managing file paths
              "VennDiagram",      # Venn Diagram
              "RColorBrewer",     # Plotting colors
              "plotly",           # Interactive plots
              "glue",             # Pasting text
              "DT", "kableExtra") # Tables             
pacman::p_load(char = packages)
```

```{r}
## Scripts with functions
source(here("scripts/GO_fun.R"))
source(here("scripts/report_funs.R"))
```

```{r}
## Set contrasts of interest
contrasts <- c("akh06_vir06", "akh24_vir24",
               "lvp06_vir06", "lvp24_vir24",
               "lvp06_akh06", "lvp24_akh24")
contrast_names <- c("AKH v. virgin (6 hpm)",
                    "AKH v. virgin (24 hpm)",
                    "LVP v. virgin (6 hpm)",
                    "LVP v. virgin (24 hpm)",
                    "AKH v. LVP (6 hpm)",
                    "AKH v. LVP (24 hpm)")

## Define tissues
tissues <- c("ab", "ht", "lrt", "lrt10")
```

```{r}
## Define input files
indir_DE <- here("results", "DE")
DE_file <- here(indir_DE, "DE_all.txt")
gene_file <- here(indir_DE, "gene_info.txt")

indir_GO <- here("results", "GO")
indir_kegg <- here("results", "kegg")
GO_res_file <- here(indir_GO, "GO_all.txt")
kegg_res_file <- here(indir_kegg, "kegg_all.txt")
```

```{r}
## Read input files 1
gene_df <- read_tsv(gene_file, show_col_types = FALSE) %>%
  select(gene_id, description) %>%
  mutate(description = str_trunc(description, width = 50))

DE_res <- read_tsv(DE_file, show_col_types = FALSE) %>%
  mutate(tissue = gsub(".*\\d(\\w+)", "\\1", treat_a),
         tissue = ifelse(sugar == "sug10", "lrt10", tissue),
         trt_a = gsub("(.*\\d+)\\w+", "\\1", treat_a),
         trt_b = gsub("(.*\\d+)\\w+", "\\1", treat_b),
         contrast = paste0(trt_a, "_", trt_b),
         contrast_full = paste0(contrast, "_", tissue),
         sig = ifelse(padj < 0.05, TRUE, FALSE)) %>%
  filter(contrast %in% contrasts) %>%            # Only select focal contrasts!
  select(gene_id, lfc, pvalue, padj, sig,
         trt_a, trt_b, tissue, sugar, contrast, contrast_full) %>%
  left_join(gene_df, by = "gene_id")
```

```{r}
## Read input files 2
GO_res <- read_tsv(GO_res_file, show_col_types = FALSE) %>%
  mutate(description = str_trunc(description, width = 50),
         contrast_full = contrast,
         sig = ifelse(padj < 0.05 & numDEInCat > 1, TRUE, FALSE),
         tissue = sub(".*\\d\\d_(\\w+)$", "\\1", contrast),
         contrast = sub("(.*_.*)_\\w+$", "\\1", contrast)) %>% 
  filter(contrast %in% contrasts)           # Only select focal contrasts!
GO_sig <- GO_res %>% filter(sig == TRUE)

kegg_res <- read_tsv(kegg_res_file, show_col_types = FALSE) %>%
  mutate(description = str_trunc(description, width = 50),
         contrast_full = contrast,
         sig = ifelse(padj < 0.05, TRUE, FALSE),
         tissue = sub(".*\\d\\d_(\\w+)$", "\\1", contrast),
         contrast = sub("(.*_.*)_\\w+$", "\\1", contrast)) %>% 
  filter(contrast %in% contrasts)           # Only select focal contrasts!
kegg_sig <- kegg_res %>% filter(sig == TRUE)
```

```{r}
## Make a df with all pairwise contrasts
contrast_df <- DE_res %>%
  mutate(time_a = gsub(".*(\\d\\d)", "\\1", trt_a),
         time_b = gsub(".*(\\d\\d)", "\\1", trt_b),
         time_contrast = factor(paste0(time_a, "_", time_b),
                                levels = c("06_06", "24_24"))) %>%
  distinct(contrast_full, .keep_all = TRUE) %>% 
  select(contrast_full, contrast, trt_a, trt_b, tissue, sugar, time_contrast)
```

<br>

-----

## Gene Ontology

### Methods

The R/Bioconductor package
[`GOseq`](https://bioconductor.org/packages/release/bioc/html/goseq.html)
was used to test for GO term enrichment among differentially expressed genes
for each pairwise contrast.
Gene Ontology assignments to genes were extracted from a VectorBase GAF
file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gaf/).
Gene lengths are used in the statistical model of `GOseq`,
and there were extracted from a VectorBase GFF file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gff/data/).
P-values were corrected for multiple testing using the Benjamini-Hochberg
correction.

### Nr significant GO terms per contrast

<br>

```{r}
sig_by_comp <- GO_sig %>%
  group_by(contrast, tissue, .drop = FALSE) %>%
  count(name = "nsig") %>%
  ungroup() %>%
  pivot_wider(id_cols = contrast, names_from = tissue, values_from = nsig) %>%
  left_join(distinct(contrast_df, contrast, .keep_all = TRUE),
            by = "contrast") %>% 
  arrange(time_contrast, trt_a, trt_b) %>%
  select(trt_a, trt_b, ab, ht, lrt, lrt10)

sig_by_comp %>%
  rename("group A" = trt_a, "group B" = trt_b,
         abdomen = ab, "head & thorax" = ht, LRT3 = lrt, LRT10 = lrt10) %>%
  make_dt(simple_mode = TRUE)
```

### Nr significant contrasts per GO term {.tabset .tabset-fade .tabset-pills}

```{r}
sig_by_term <- function(GO_df, tissues) {
  
  if (length(tissues) > 1) GO_df <- GO_df %>% mutate(contrast = contrast_full) 
  
  GO_df %>%
    filter(tissue %in% tissues) %>%
    add_count(category) %>%
    group_by(category) %>%
    mutate("signif. contrasts" = paste(contrast, collapse = ", ")) %>%
    select("n significant" = n, category, ontology, description,
           `signif. contrasts`) %>%
    make_dt()
}
```

#### Abdomen

<br>

```{r}
sig_by_term(GO_sig, tissues = "ab")
```

#### Head & thorax

<br>

```{r}
sig_by_term(GO_sig, tissues = "ht")
```

#### LRT (3%)

<br>

```{r}
sig_by_term(GO_sig, tissues = "lrt")
```

#### LRT 10%

<br>

```{r}
sig_by_term(GO_sig, tissues = "lrt10")
```

#### All tissues

<br>

```{r}
sig_by_term(GO_sig, tissues = c("ab", "ht", "lrt", "lrt10"))
```

### All significant GO terms

```{r}
sig_smr <- pivot_wider(GO_sig,
                       id_cols = c(category, description),
                       names_from = contrast_full,
                       values_from = padj)

numr_cols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]
make_dt(sig_smr, numr_cols = numr_cols)
```

### Heatmaps {.tabset .tabset-fade .tabset-pills}

Gray squares indicate non-significant values.

#### Abdomen

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "ab", contrasts) %>%
  goplot(ylabsize = 7)
```

#### Head & thorax

```{r, out.width="100%"}
prep_goplot(GO_res, tissues = "ht", contrasts) %>%
  goplot()
```

#### LRT (3%)

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt", contrasts) %>%
  goplot(ylabsize = 7)
```

#### LRT 10%

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt10", contrasts) %>% 
  goplot(ylabsize = 7)
```

#### All tissues

```{r, fig.height = 8, out.width="100%"}
prep_goplot(GO_res, tissues = tissues, contrasts) %>%
  goplot(ylabsize = 5)
```

<br>

----

## KEGG Ontology

### Methods

The R/Bioconductor package
[`clusterProfiler`](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
was used to test for KEGG term enrichment among differentially expressed genes
for each pairwise contrast.
KEGG pathway assignments to genes were obtained using the R/Bioconductor package 
[`KEGGREST`](https://bioconductor.org/packages/release/bioc/html/KEGGREST.html),
since KEGG mappings are available for _A. aegypti_.
P-values were corrected for multiple testing using the Benjamini-Hochberg
correction.

### Nr significant pathways per contrast

```{r}
sig_by_comp <- kegg_sig %>%
  group_by(contrast, tissue, .drop = FALSE) %>%
  count(name = "nsig") %>%
  ungroup() %>%
  pivot_wider(id_cols = contrast, names_from = tissue, values_from = nsig) %>%
    left_join(distinct(contrast_df, contrast, .keep_all = TRUE),
            by = "contrast") %>%
  arrange(time_contrast, trt_a, trt_b) %>%
  select(trt_a, trt_b, ab, ht, lrt, lrt10)

sig_by_comp %>%
  rename("group A" = trt_a, "group B" = trt_b,
         abdomen = ab, "head & thorax" = ht, LRT3 = lrt, LRT10 = lrt10) %>%
  make_dt(simple_mode = TRUE)
```

### Nr significant contrasts per pathway {.tabset .tabset-fade .tabset-pills}

```{r}
sig_by_pathway <- function(kegg_df, tissues) {
  
  if (length(tissues) > 1) kegg_df <- kegg_df %>% mutate(contrast = contrast_full) 
  
  kegg_df %>%
    filter(tissue %in% tissues) %>%
    add_count(category) %>%
    group_by(category) %>%
    mutate("signif. contrasts" = paste(contrast, collapse = ", ")) %>%
    select("n significant" = n, pathway = category, description,
           `signif. contrasts`) %>%
    make_dt()
}
```

#### Abdomen

<br>
```{r}
sig_by_pathway(kegg_sig, tissues = "ab")
```

#### Head & thorax

<br>

```{r}
sig_by_pathway(kegg_sig, tissues = "ht")
```

#### LRT (3%)

<br>

```{r}
sig_by_pathway(kegg_sig, tissues = "lrt")
```

#### LRT 10%

<br>

```{r}
sig_by_pathway(kegg_sig, tissues = "lrt10")
```

#### All tissues

```{r}
sig_by_pathway(kegg_sig, tissues = c("ab", "ht", "lrt", "lrt10"))
```

```{r}
sig_by_term <- kegg_sig %>%
  add_count(category) %>%
  group_by(category) %>%
  mutate("signif. contrasts" = paste(contrast, collapse = ", ")) %>%
  select("n significant" = n, pathway = category, description, `signif. contrasts`)

make_dt(sig_by_term)
```

### All significant pathways

<br>

```{r}
sig_smr <- pivot_wider(kegg_sig,
                       id_cols = c(category, description),
                       names_from = contrast_full,
                       values_from = padj)

numr_cols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]

make_dt(sig_smr, numr_cols = numr_cols)
```

### Heatmaps {.tabset .tabset-fade .tabset-pills}

Gray squares indicate non-significant values.

#### Abdomen

```{r, fig.height=5, out.width="100%"}
prep_goplot(kegg_res, tissues = "ab", contrasts) %>%
  goplot()
```

#### Head & thorax

```{r, fig.height=5, out.width="100%"}
prep_goplot(kegg_res, tissues = "ht", contrasts) %>%
  goplot()
```

#### LRT (3%)

```{r, fig.height=5, out.width="100%"}
prep_goplot(kegg_res, tissues = "lrt", contrasts) %>%
  goplot()
```

#### LRT 10%

```{r, fig.height=5, out.width="100%"}
prep_goplot(kegg_res, tissues = "lrt10", contrasts) %>%
  goplot()
```

#### All tissues

```{r, fig.height=5, out.width="100%"}
prep_goplot(kegg_res, tissues = tissues, contrasts) %>%
  goplot(ylabsize = 7)
```
