---
title: |
  | _Aedes aegypti_ mRNA expression --
  | Gene ontology analyses
pagetitle: "Enrichment results"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster, OSU"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
## Install/load packages
if(! "pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse",        # Misc. data manipulation and plotting
              "here",             # Managing file paths
              "VennDiagram",      # Venn Diagram
              "ggpubr",           # Cleveland dot plots
              "ggforce",          # facet_col for equal tile heights
              "RColorBrewer",     # Plotting colors
              "plotly",           # Interactive plots
              "glue",             # Pasting text
              "DT", "kableExtra") # Tables             
pacman::p_load(char = packages)
```

```{r}
## Scripts with functions
source(here("scripts/GO_fun.R"))
source(here("scripts/report_funs.R"))
```

```{r}
## Set contrasts of interest
contrasts <- c("lvp06_vir06", "lvp24_vir24")
contrast_labs <- c("6 hpm", "24 hpm")
contrasts_full <- c("lvp06_vir06_ab", "lvp24_vir24_ab",
                    "lvp06_vir06_ht", "lvp24_vir24_ht",
                    "lvp06_vir06_lrt", "lvp24_vir24_lrt",
                    "lvp06_vir06_lrt12", "lvp24_vir24_lrt12")
  
## Define tissues
tissues <- c("ab", "ht", "lrt", "lrt12")
```

```{r}
## Define input files
indir_DE <- here("results", "DE")
DE_file <- here(indir_DE, "DE_all.txt")
gene_file <- here(indir_DE, "gene_info.txt")

indir_GO <- here("results", "GO")
GO_res_file <- here(indir_GO, "GO_all.txt")
GO_map_file <- here(indir_GO, "GO_map.txt")
```

```{r}
## Read input files 1
gene_df <- read_tsv(gene_file, show_col_types = FALSE) %>%
  select(gene_id, description) %>%
  mutate(description = str_trunc(description, width = 50))

DE_res <- read_tsv(DE_file, show_col_types = FALSE) %>%
  mutate(tissue = gsub(".*\\d(\\w+)", "\\1", treat_a),
         tissue = ifelse(sugar == "sug12", "lrt12", tissue),
         trt_a = gsub("(.*\\d+)\\w+", "\\1", treat_a),
         trt_b = gsub("(.*\\d+)\\w+", "\\1", treat_b),
         contrast = paste0(trt_a, "_", trt_b),
         contrast_full = paste0(contrast, "_", tissue),
         sig = ifelse(padj < 0.05, TRUE, FALSE)) %>%
  filter(contrast %in% contrasts) %>%            # Only select focal contrasts!
  select(gene_id, lfc, pvalue, padj, sig,
         trt_a, trt_b, tissue, sugar, contrast, contrast_full) %>%
  left_join(gene_df, by = "gene_id")
```

```{r}
## Read input files 2
GO_map <- read_tsv(GO_map_file, show_col_types = FALSE)

GO_res <- read_tsv(GO_res_file, show_col_types = FALSE) %>%
  filter(category != "GO:0055114") %>%  # Remove obsolete GO category
  mutate(description = str_trunc(description, width = 50),
         contrast_full = contrast,
         sig = ifelse(padj < 0.05 & numDEInCat > 1, TRUE, FALSE),
         tissue = sub(".*\\d\\d_(\\w+)$", "\\1", contrast),
         contrast = sub("(.*_.*)_\\w+$", "\\1", contrast)) %>% 
  filter(contrast %in% contrasts)           # Only select focal contrasts!

GO_sig <- GO_res %>% filter(sig == TRUE)

GO_lookup <- GO_res %>%
  select(category, ontology, description) %>% 
  distinct()
```

```{r}
## Make a df with all pairwise contrasts
contrast_df <- DE_res %>%
  mutate(time_a = gsub(".*(\\d\\d)", "\\1", trt_a),
         time_b = gsub(".*(\\d\\d)", "\\1", trt_b),
         time_contrast = factor(paste0(time_a, "_", time_b),
                                levels = c("06_06", "24_24"))) %>%
  distinct(contrast_full, .keep_all = TRUE) %>% 
  select(contrast_full, contrast, trt_a, trt_b, tissue, sugar, time_contrast)
```

<br>

-----

## Methods

The R/Bioconductor package
[`GOseq`](https://bioconductor.org/packages/release/bioc/html/goseq.html)
was used to test for GO term enrichment among differentially expressed genes
for each pairwise contrast.
Gene Ontology assignments to genes were extracted from a VectorBase GAF
file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gaf/).
Gene lengths are used in the statistical model of `GOseq`,
and there were extracted from a VectorBase GFF file available
[here](https://vectorbase.org/vectorbase/app/downloads/release-49/AaegyptiLVP_AGWG/gff/data/).
P-values were corrected for multiple testing using the Benjamini-Hochberg
correction.

**Added notes 2021-02-19**
- GO category `GO:0055114`, which was present in earlier versions of this document,
  was removed because it is an obsolete category.
  This category was related to oxidation-reduction which is now reflected by
  `GO:0016491` (oxidoreductase activity).
  
**Added notes 2021-04-24**  
- In some cases, a GO category is nominally significant even though only a single
  gene in the category is DE.
  I am now requiring that there are at least 2 DEGs
  (in addition to an adjusted p-value below 0.05) for a GO category to be significant.


<br>

-----

## Nr significant GO terms per contrast

<br>

```{r}
sig_by_comp <- GO_sig %>%
  group_by(contrast, tissue, direction, .drop = FALSE) %>%
  count(name = "nsig") %>%
  ungroup() %>%
  pivot_wider(id_cols = c(contrast, direction),
              names_from = tissue, values_from = nsig,
              values_fill = 0) %>%
  left_join(distinct(contrast_df, contrast, .keep_all = TRUE),
            by = "contrast") %>% 
  arrange(time_contrast, trt_a, trt_b) %>%
  select(trt_a, trt_b, direction, ab, ht, lrt, lrt12)

sig_by_comp %>%
  rename("group A" = trt_a, "group B" = trt_b,
         "Ab" = ab, "HT" = ht, "LRT3" = lrt, "LRT12" = lrt12) %>%
  make_kable()
```


<br>

-----

## All significant GO terms {.tabset .tabset-fade .tabset-pills}

<br>

### Both DE directions

<br>

```{r}
sig_smr <- pivot_wider(GO_sig %>% filter(direction == "both"),
                       id_cols = c(category, description),
                       names_from = contrast_full,
                       values_from = padj)

numr_cols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]
make_dt(sig_smr, numr_cols = numr_cols)
```

<br>

-----

### Upregulated after mating

<br>

```{r}
sig_smr <- pivot_wider(GO_sig %>% filter(direction == "up"),
                       id_cols = c(category, description),
                       names_from = contrast_full,
                       values_from = padj)

numr_cols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]
make_dt(sig_smr, numr_cols = numr_cols)
```

<br>

-----

### Downregulated after mating

<br>

```{r}
sig_smr <- pivot_wider(GO_sig %>% filter(direction == "down"),
                       id_cols = c(category, description),
                       names_from = contrast_full,
                       values_from = padj)

numr_cols <- colnames(sig_smr)[str_detect(colnames(sig_smr), "_")]
make_dt(sig_smr, numr_cols = numr_cols)
```

<br>

-----

## All DEGS in significant GO terms {.tabset .tabset-fade .tabset-pills}

<br>

```{r}
get_go_degs <- function(fcontrast, direction,
                        GO_map, GO_sig, DE_res, gene_df) {
  
  fGO_sig <- GO_sig %>%
    filter(contrast_full == fcontrast,
           direction == direction) %>%
    select(-sig, -contrast_full) %>%
    rename(description_GO = description)
  
  DEGs <- DE_res %>%
    filter(contrast_full == fcontrast,
           padj < 0.05) %>%
    pull(gene_id)
  
  GO_sig_wgenes <- GO_map %>% 
    filter(gene_id %in% DEGs,
           go_term %in% fGO_sig$category) %>%
    full_join(fGO_sig, ., by = c("category" = "go_term")) %>% 
    left_join(gene_df, by = "gene_id") %>%
    rename(description_gene = description) %>% 
    select(-DE_in_cat, -numInCat) %>% 
    relocate(tissue, .after = contrast)
}
```

### Both DE directions

<br>

```{r}
go_degs <- map_dfr(contrasts_full, get_go_degs,
                   direction = "both",
                   GO_map, GO_sig, DE_res, gene_df)
make_dt(go_degs)
```

<br>

-----

### Upregulated after mating

<br>

```{r}
go_degs <- map_dfr(contrasts_full, get_go_degs,
                   direction = "up",
                   GO_map, GO_sig, DE_res, gene_df)
make_dt(go_degs)
```

<br>

-----

#### Downregulated after mating

```{r}
go_degs <- map_dfr(contrasts_full, get_go_degs,
                   direction = "down",
                   GO_map, GO_sig, DE_res, gene_df)
make_dt(go_degs)
```

<br>

-----

## Heatmaps - both DE directions {.tabset .tabset-fade .tabset-pills}

<br>

- **For DEGs in either direction (over- or underexpressed after mating).**

- Gray squares indicate non-significant values.
- Numbers indicate the number of DEGs in the focal category.
- If a timepopint is missing, there were no significant categories.

<br>

### Abdomen

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "ab", contrasts) %>%
  goplot(ylabsize = 7) +
  scale_x_discrete(labels = contrast_labs)
```

<br>

-----

### Head & thorax

```{r, out.width="100%"}
prep_goplot(GO_res, tissues = "ht", contrasts) %>%
  goplot() +
  scale_x_discrete(labels = contrast_labs)
```

<br>

-----

### LRT (3%)

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt", contrasts) %>%
  goplot(ylabsize = 7) +
  scale_x_discrete(labels = contrast_labs)
```

<br>

-----

### LRT 12%

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt12", contrasts) %>% 
  goplot(ylabsize = 7) +
  scale_x_discrete(labels = contrast_labs)
```

```{r, fig.height = 8, out.width="100%"}
# #### All tissues
#prep_goplot(GO_res, tissues = tissues, contrasts) %>%
#   goplot(ylabsize = 5)
```

<br>

-----

## Heatmaps - separate DE directions {.tabset .tabset-fade .tabset-pills}

<br>

**These plots have separate entries for enriched GO categories for down- and
upregulated genes after mating.**

- Gray squares indicate non-significant values.
- Numbers indicate the number of DEGs in the focal category.
- If a timepopint is missing, there were no significant categories.

<br>

### Abdomen

If omitted the DEG count labels for this plot because they were too squished
due to the high number of GO categories.

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "ab", contrasts = contrasts,
            directions = c("up", "down")) %>%
  complete(category, contrast, direction) %>% 
  select(-ontology, -description) %>%
  left_join(GO_lookup, by = "category") %>% 
  mutate(contrast = ifelse(contrast == "lvp06_vir06", "06 hpm", "24 hpm")) %>% 
  goplot(x_var = "direction", facet_var = "contrast",
         label_count = FALSE, ylabsize = 7)
```

### Head & thorax

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "ht", contrasts = contrasts,
            directions = c("up", "down")) %>%
  complete(category, contrast, direction) %>% 
  select(-ontology, -description) %>%
  left_join(GO_lookup, by = "category") %>% 
  mutate(contrast = ifelse(contrast == "lvp06_vir06", "06 hpm", "24 hpm")) %>% 
  goplot(x_var = "direction", facet_var = "contrast", ylabsize = 7)
```

### LRT 3%

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt", contrasts = contrasts,
            directions = c("up", "down")) %>%
  complete(category, contrast, direction) %>% 
  select(-ontology, -description) %>%
  left_join(GO_lookup, by = "category") %>% 
  mutate(contrast = ifelse(contrast == "lvp06_vir06", "06 hpm", "24 hpm")) %>% 
  goplot(x_var = "direction", facet_var = "contrast", ylabsize = 7)
```

### LRT 12%

```{r, fig.height = 6, out.width="100%"}
prep_goplot(GO_res, tissues = "lrt12", contrasts = contrasts,
            directions = c("up", "down")) %>%
  complete(category, contrast, direction) %>% 
  select(-ontology, -description) %>%
  left_join(GO_lookup, by = "category") %>% 
  mutate(contrast = ifelse(contrast == "lvp06_vir06", "06 hpm", "24 hpm")) %>% 
  goplot(x_var = "direction", facet_var = "contrast", ylabsize = 7)
```

<br>